### DISCUSSION TITLE
Nested Transactions

### DISCUSSION THREAD

**Discussion by i-zanev** [nhibernate/nhibernate-core#2880] [Q&A]
Hi Everyone,

I've been trying to implement nested transactions with NHibernate using Sybase but to no luck really. 
I'll try to explain my use case: 

>Begin transaction
>
>            Begin transaction
>                        Do some work
>            Commit
>
>            Begin transaction
>                        Do some more work
>            Commit
>
>If error Rollback(and that would rollback both inner transactions)
>Else commit(that commits the changes of both transactions)

It seems like no matter if I rollback or commit my outer transaction the changes by the inner ones are always persisted. 

Many thanks for any feedback!

Thanks,
Iliya

**Comment by fredericDelaporte** [comment]
Nested transactions are not supported. See [here](https://stackoverflow.com/a/3363823/1178314).

**Reply by i-zanev** [reply]
Thanks fredericDelaporte. I did see that post earlier on but since it's not that recent I wanted to check if things haven't changed. 
That post also suggested that you can achieve the same outcome if you wrap transactions in a TransactionScope but I tried that and it didn't work for me. Not sure if there's a cfg option or something that I'm missing...

**Reply by fredericDelaporte** [reply]
Do not use transactions if you use transaction scopes. Those two transactions API are not compatible. NHibernate transactions use DbTransaction, and Microsoft documents somewhere that DbTransaction should not be used with transaction scopes.

You may nest transaction scopes, but unless you open sessions only inside the nested transaction scopes, it may be a challenge to get the session enlisted in the right scope. And I am not sure anyway how behave nested scopes, what it will imply for NHibernate sessions, can it even work...

**Reply by i-zanev** [reply]
Thanks fredericDelaporte. I think I've managed to get the desired behaviour using TransactionScopes and also fitting that within the way the API works at the moment :)

**Reply by fredericDelaporte** [reply]
See also the [NHibernate documentation about transaction scope](https://nhibernate.info/doc/nhibernate-reference/transactions.html#transactions-scopes):
>  When using `TransactionScope` with NHibernate, you need to be aware of following points:
>
>  - The session will enlist with the first scope in which the session is used (or opened). As of NHibernate v5.0, it will enlist its connection in the transaction regardless of connection string `Enlist` setting. Prior to v5.0, it was relying on that setting being considered `true`, and on acquiring the connection within the scope.
>
>    Sub-scopes are not supported. The session will be enlisted in the first scope within which it was used, until this scope is committed or rollback. If auto-enlistment is enabled on the connection and the session used on others scopes than the one in which it is currently enlisted, the connection may enlist in another scope, and the session will then fail to use it.
>
>    Session auto-enlistment can be controlled from the session builder obtained with `ISessionFactory.WithOptions()`, using the `AutoJoinTransaction` option. It can also be controlled at the configuration level, see `transaction.auto_join` in [Section 3.6, “Optional configuration properties”](https://nhibernate.info/doc/nhibernate-reference/session-configuration.html#configuration-optional). When auto-join is disabled, the connection may still enlist itself if connection string `Enlist` setting is not false. A session can explicitly join the current system transaction by calling `ISession.JoinTransaction()`.
